Sources/SwiftParser/Declarations.swift

```diff
@@ -311,6 +325,185 @@ extension Parser {
     )
   }
 
+  /// Parse the standalone header of a declaration group (a header that won't have a member block attached to it).
+  /// For a header that will have a member list attached, use ``Parser/parseHeaderForDeclarationGroup(attrs:keyword:handle:)``, which this method wraps.
+  ///
+  /// If `inMemberDeclList` is `true`, we know that the next item must be a
+  /// declaration and thus start with a keyword. This allows further recovery.
+  mutating func parseDeclarationGroupHeader(inMemberDeclList: Bool = false) -> RawDeclGroupHeaderSyntax {
+    // If we are at a `#if` of attributes, the `#if` directive should be
+    // parsed when we're parsing the attributes.
+    if self.at(.poundIf) && !self.withLookahead({ $0.consumeIfConfigOfAttributes() }) {
+      // If we're at a #if, that's not actually allowed here; parse it and add
+      // it as an unexpected node of a MissingDeclGroupHeaderSyntax.
+      let directive = self.parsePoundIfDirective { (parser, _) in
+        let parsedDecl = parser.parseDeclaration()
+        let semicolon = parser.consume(if: .semicolon)
+        return RawMemberBlockItemSyntax(
+          decl: parsedDecl,
+          semicolon: semicolon,
+          arena: parser.arena
+        )
+      } addSemicolonIfNeeded: { lastElement, newItemAtStartOfLine, parser in
+        if lastElement.semicolon == nil && !newItemAtStartOfLine {
+          return RawMemberBlockItemSyntax(
+            lastElement.unexpectedBeforeDecl,
+            decl: lastElement.decl,
+            lastElement.unexpectedBetweenDeclAndSemicolon,
+            semicolon: parser.missingToken(.semicolon),
+            lastElement.unexpectedAfterSemicolon,
+            arena: parser.arena
+          )
+        } else {
+          return nil
+        }
+      } syntax: { parser, elements in
+        return .decls(RawMemberBlockItemListSyntax(elements: elements, arena: parser.arena))
+      }
+      return RawDeclGroupHeaderSyntax(
+        RawMissingDeclHeaderSyntax(
+          RawUnexpectedNodesSyntax([directive], arena: self.arena),
+          attributes: emptyCollection(RawAttributeListSyntax.self),
+          modifiers: emptyCollection(RawDeclModifierListSyntax.self),
+          placeholder: missingToken(.identifier, text: "decl"),
+          inheritanceClause: nil,
+          genericWhereClause: nil,
+          arena: self.arena
+        )
+      )
+    }
+
+    let attrs = DeclAttributes(
+      attributes: self.parseAttributeList(),
+      modifiers: self.parseDeclModifierList()
+    )
+
+    let recoveryResult: (match: DeclGroupHeaderSyntax.IntroducerOptions, handle: RecoveryConsumptionHandle)?
+    if let atResult = self.at(anyIn: DeclGroupHeaderSyntax.IntroducerOptions.self) {
```

Comment:
Why can’t we always use `self.canRecoverTo`? That should already contain the case for if we are at a `DeclGroupHeaderSyntax.IntroducerOptions.self`.

Reply:
I must admit that I based this on `parseDeclaration()` and didn't look closely at whether this method could be written to always use `canRecoverTo`. If you think it can, I'll try to modify it.

Reply:
Yes, I think `canRecoverTo` should work.